devlib/utils/android.py

```diff
@@ -701,19 +632,18 @@ def wait_for(self, regexp, timeout=30):
         :returns: List of matched strings
         """
         res = self.search(regexp)
-
         # Found some matches, return them
-        # Also return if thread not running
-        if len(res) > 0 or not self.is_alive():
+        if len(res) > 0:
             return res
 
-        # Did not find any match, wait for one to pop up
-        self._sought = regexp
-        found = self._match_found.wait(timeout)
-        self._match_found.clear()
-        self._sought = None
+        # Store the number of lines we've searched already, so we don't have to
+        # re-grep them after 'expect' returns
+        next_line_num = len(self.get_log())
```

Comment:
Mmm unlikely but not impossible: between the `if len(res) > 0` and this index store, we could've received a matching line. That's something the `_datalock` was taking care of in the previous implementation. Could it be possible to freeze the `pexpect` process before doing the `search`, unfreeze it if we found lines or after we've saved the current index ?

Reply:
That isn't a problem - `expect` still checks lines that happened before it got called.

Reply:
Oh, right, didn't read that properly. Looks good to me !

Reply:
Oh yeah damn you're right. I'll fix it by gluing together the `self.search()` and `self.get_log()`

Reply:
OK, should be fixed now. Thanks a lot for the careful review :)

Reply:
I think you could just store `next_line_num` before doing a call to `search` - which I believe is equivalent to what you're doing now (?)

Reply:
Mm actually it might just be better to keep what you have right now but do
```
res = [line for line in log if re.match(regexp, line)]
```
instead of
```
res = [line for line in self.get_log() if re.match(regexp, line)]
```

Reply:
OH damn sorry that's exactly what I meant to do, just rushed it. Updating once again.